### PR TITLE
homework 2 done

### PR DESCRIPTION
**Pull request by PIDAMI**

### CODE REVIEW

File: src/test/java/com/epam/tc/hw2/BaseTest.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.BeforeTest;
+
+public class BaseTest {
+    protected WebDriver driver;
+    private static final long TIMEOUT_SECONDS = 10;
+    protected final String indexPageURL = "https://jdi-testing.github.io/jdi-light/index.html";
+    protected WebDriverWait wait;
+    protected final User user = loadUserFromProperties();
+    protected final String expectedBrowserTitle = "Home Page";
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/BaseTest.java]: All final constants should have name in upper case. User may be not final.

File: src/test/java/com/epam/tc/hw2/BaseTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+
+    // 1. Open test site by URL
+    public void gotoSite() {
+        if (driver == null) {
+            System.out.println("driver is null :(");
+            driver = new ChromeDriver();
+        }
+        driver.get(indexPageURL);
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/BaseTest.java]: If driver was null for some reason, you need just throw an exception. In your case you set up new ChromeDriver, but don't do next steps. And 'WebDriverWait wait' will be null.
driver.manage().window().maximize();
wait = new WebDriverWait(driver, Duration.ofSeconds(TIMEOUT_SECONDS));

File: src/test/java/com/epam/tc/hw2/BaseTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        WebElement loginButton = driver.findElement(By.id("login-button"));
+        loginButton = wait.until(ExpectedConditions.elementToBeClickable(loginButton));
+        loginButton.click();
+    }
+
+    // 4. Assert Username is loggined;
+    public void checkUserIsLoggined(User user) {
+        WebElement username = driver.findElement(By.cssSelector("span#user-name"));
+        assertThat(username.isDisplayed()).isEqualTo(true);
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/BaseTest.java]: Can be simplified to username.isDisplayed();

File: src/test/java/com/epam/tc/hw2/BaseTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            String password = prop.getProperty("password");
+            String displayedName = prop.getProperty("displayedName");
+
+            return new User(username, password, displayedName);
+
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/BaseTest.java]: Here should be collapsed to second catch. Second catch identical to the first.

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    private final List<String> expectedHeaderButtonsText = Stream
+        .of("HOME", "CONTACT FORM", "SERVICE", "METALS & COLORS")
+        .sorted()
+        .collect(Collectors.toList());
+
+    private final String frameButtonValue = "Frame Button";
+    private final List<String> expectedLeftSectionItemsText = Stream
+        .of("Home", "Contact form", "Service", "Metals & Colors", "Elements packs")
+        .sorted()
+        .collect(Collectors.toList());
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: All final constants should have name in upper case.
All test data can be extended with data class.

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+
+
+
+
+    @Test
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Please, leave only one row between methods or class fields.

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        );
+        assertThat(headerNavigationButtons.size()).isEqualTo(expectedHeaderButtonsText.size());
+        List<String> headerButtonsActualText = headerNavigationButtons
+            .stream()
+            .map(WebElement::getText)
+            .sorted()
+            .collect(Collectors.toList());
+
+        assertThat(headerButtonsActualText).isEqualTo(expectedHeaderButtonsText);
+        headerNavigationButtons.forEach(el -> assertThat(el.isDisplayed()).isEqualTo(true));
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Again should be simplified.

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        checkLeftSectionHasItemsWithProperText(expectedLeftSectionItemsText);
+    }
+
+
+
+    // 5. Assert that there are 4 items on the header section
+    // are displayed and they have proper texts;
+    public void checkAllHeadersItemsDisplayed(List<String> expectedHeaderButtonsText) {
+        List<WebElement> headerNavigationButtons = driver.findElements(
+            By.cssSelector(".uui-navigation.nav.navbar-nav > *")
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Weak locator. Can be modified to more stable.

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                String.format("//*[@value='%s']", frameButtonValue))
+            );
+            int numButtons = frameButtons.size();
+            switchToOriginalWindow();
+            if (numButtons > 0) {
+                frameWithButtonExists = true;
+                break;
+            }
+        }
+        assertThat(frameWithButtonExists).isEqualTo(true);
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Here can be like that assertThat(frameWithButtonExists).isTrue();

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+    // 10. Switch to original window back;
+    // used in checkFrameWithButtonExists multiple times
+    // to switch from frame to original window
+    public void switchToOriginalWindow() {
+        driver.switchTo().defaultContent();
+    }
+
+    public WebElement getTextNodeFromImageNode(WebElement image) {
+        WebElement parent = image.findElement(By.xpath("./../.."));
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Such locator hard understandable in future. Only you know what does it mean. Can it be changed on clear one?

File: src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        checkboxes.stream()
+                  .filter(el -> checkboxTextToLogFormat.containsKey(el.getText()))
+                  .forEach(el -> {
+                      // log is element with text such as listed in map; assert that this log exists
+                      String logRowFormat = checkboxTextToLogFormat.get(el.getText());
+
+                      List<WebElement> logRows = driver.findElements(By.xpath(
+                          String.format("//*[contains(text(),'%s')]", logRowFormat))
+                      );
+                      assertThat(logRows.size()).isNotEqualTo(0);
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java]: Use .isNotZero()

**Comment by Delnote** [comment]
Some common comments:

- leave one row between methods and class fields
- store data in data classes, axillary methods in util or other axillary classes. Test class can have (Before, After if they are not in Base class), Test methods.
- Replace all .isEqualTo(true)

File: src/test/java/com/epam/tc/hw2/BaseTest.java (hunk 1 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/BaseTest.java]: Made url and browser title into static final constants with names in Upper case. User field removed and data provider class created, user is now loaded in its dataProvide method

File: src/test/java/com/epam/tc/hw2/BaseTest.java (hunk 2 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/BaseTest.java]: Removed checks for null

File: src/test/java/com/epam/tc/hw2/BaseTest.java (hunk 3 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/BaseTest.java]: Done

File: src/test/java/com/epam/tc/hw2/BaseTest.java (hunk 4 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/BaseTest.java]: Collapsed to one catch-block

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java (hunk 5 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Moved to dataProvide method of corresponding class

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java (hunk 6 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Removed unnecessary blank rows

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java (hunk 8 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Modified to `nav > ul.uui-navigation.nav.navbar-nav.m-l8 > li`

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java (hunk 7 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Done

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java (hunk 9 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Done

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java (hunk 10 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseTest.java]: Removed this util method. Elements with text after images are now explicitly found with `div.row.clerafix.benefits > div.col-sm-3 > div.benefit > span.benefit-txt` locators.

File: src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java (hunk 11 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java]: Done

File: src/test/java/com/epam/tc/hw2/BaseTest.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.BeforeSuite;
+
+public class BaseTest {
+    protected WebDriver driver;
+    private static final long TIMEOUT_SECONDS = 10;
+    protected WebDriverWait wait;
+    protected BaseData baseData;
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/BaseTest.java]: Unused class field

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            .sorted()
+            .collect(Collectors.toList());
+
+        String frameButtonValue = "Frame Button";
+
+        return new Object[][]{
+            {INDEX_PAGE_URL, EXPECTED_BROWSER_TITLE, user,
+                imageCssLocatorToTextAfterImage, expectedHeaderButtonsText,
+                expectedLeftSectionItemsText, frameButtonValue}
+        };
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java]: Data provider is never using in such way. In your case you need just to have your tests data as it was previously, but in separate data class. It's needed to be separated: tests in test classes, it's data in data (factory) classes.

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+
+
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java:48]: Please remove empty rows in all classes. Should be only one empty row between methods or fields.

File: src/test/java/com/epam/tc/hw2/ex2/SecondExerciseData.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        );
+
+        Map<String, String> colorTextToLogFormat = Map.ofEntries(
+            entry("Yellow", "Colors: value changed to Yellow")
+        );
+
+        return new Object[][]{
+            {INDEX_PAGE_URL, EXPECTED_BROWSER_TITLE, user, checkboxElementsTextToLogFormat,
+                checkboxRadioTextToLogFormat, colorTextToLogFormat}
+        };
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex2/SecondExerciseData.java]: The same for dataProvider.

File: src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.Select;
+import org.testng.annotations.Test;
+
+public class SecondExerciseTest extends BaseTest {
+
+    // path not to input node but to its parent which has text
+    public static final String CHECKBOX_ELEMENTS_XPATH = "//label[@class='label-checkbox']";
+    public static final String CHECKBOX_RADIO_XPATH = "//label[@class='label-radio']";
+    public static final String COLOR_XPATH = "//div[@class='colors']/*";
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java]: Locators can't be a part of test class. It should be moved in Tests or inside methods.

File: src/test/java/com/epam/tc/hw2/BaseTest.java (hunk 12 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/BaseTest.java]: Removed

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java (hunk 13 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java]: Removed dataProvide method, added separate getters for each field

File: src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java (hunk 14 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex1/FirstExerciseData.java:48]: Done

File: src/test/java/com/epam/tc/hw2/ex2/SecondExerciseData.java (hunk 15 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex2/SecondExerciseData.java]: Done

File: src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java (hunk 16 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw2/ex2/SecondExerciseTest.java]: Moved to utility class

**Comment by Delnote** [comment]
Done!

For the future, if you use data factory of some entities or just data fields in special class, it's not needed to extend them with getters. Just make them static.

**Pull request merged**